sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+
+  mqtt5->_internal.options = options == NULL ? az_mqtt5_options_default() : *options;
+  mqtt5->_internal.pahoasync_handle = pahoasync_handle;
+  mqtt5->_internal.platform_mqtt5.pipeline = NULL;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_mqtt5_outbound_connect(az_mqtt5* mqtt5, az_mqtt5_connect_data* connect_data)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt5);
+  _az_PRECONDITION_NOT_NULL(connect_data);
+  _az_PRECONDITION_VALID_SPAN(connect_data->host, 1, false);
+  _az_PRECONDITION_VALID_SPAN(connect_data->client_id, 1, false);
+
+  az_result ret = AZ_OK;
+  az_mqtt5* me = (az_mqtt5*)mqtt5;
+  MQTTAsync_createOptions create_opts = MQTTAsync_createOptions_initializer5;
+  MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer5;
+  MQTTAsync_SSLOptions ssl_opts = MQTTAsync_SSLOptions_initializer;
+
+  if (me->_internal.options.disable_tls == 0)
+  {
+    ssl_opts.verify = 1;
+    bool use_os_certs
+        = (az_span_ptr(me->_internal.options.certificate_authority_trusted_roots) == NULL);
+
+    if (!use_os_certs)
+    {
+      ssl_opts.disableDefaultTrustStore = 1;
+      ssl_opts.CApath
+          = (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots);
+    }
+
+    ssl_opts.keyStore = (const char*)az_span_ptr(connect_data->certificate.cert);
+    ssl_opts.privateKey = (const char*)az_span_ptr(connect_data->certificate.key);
+  }
+
+  if (az_span_ptr(connect_data->username) != NULL)
+  {
+    conn_opts.username = (const char*)az_span_ptr(connect_data->username);
+    conn_opts.password = (const char*)az_span_ptr(connect_data->password);
+  }
+
+  char port_buffer[SERVER_URI_PORT_MAX_LENGTH];
+  sprintf(port_buffer, "%d", connect_data->port);
+  az_span port_span = az_span_create_from_str(port_buffer);
+
+  int32_t server_uri_length
+      = az_span_size(connect_data->host) + SERVER_URI_COLON_LENGTH + SERVER_URI_PORT_MAX_LENGTH + 1;
+  char server_uri_buffer[server_uri_length];
+  az_span server_uri = az_span_create((uint8_t*)server_uri_buffer, server_uri_length);
+  az_span server_uri_temp = server_uri;
+
+  server_uri_temp = az_span_copy(server_uri_temp, connect_data->host);
+  server_uri_temp = az_span_copy_u8(server_uri_temp, ':');
+  server_uri_temp = az_span_copy(server_uri_temp, port_span);
+  server_uri_temp = az_span_copy_u8(server_uri_temp, '\0');
+
+  _az_RETURN_IF_FAILED(_az_result_from_pahoasync(MQTTAsync_createWithOptions(
+      me->_internal.pahoasync_handle,
+      (const char*)az_span_ptr(server_uri),
+      (const char*)az_span_ptr(connect_data->client_id),
+      MQTTCLIENT_PERSISTENCE_NONE,
+      NULL,
+      &create_opts)));
+
+  if (me->_internal.pahoasync_handle == NULL)
+  {
+    ret = AZ_ERROR_OUT_OF_MEMORY;
+    return ret;
+  }
+
+  _az_RETURN_IF_FAILED(_az_result_from_pahoasync(MQTTAsync_setCallbacks(
+      *me->_internal.pahoasync_handle,
+      mqtt5,
+      _az_pahoasync_on_disconnect,
+      _az_pahoasync_on_message,
+      NULL)));
```

Comment:
is this the puback callback, and do we want to set it?

Reply:
The puback callback is set when we publish in `az_mqtt5_outbound_pub`, I already tested setting this callback and I am unable to receive PUBACKs there